src/criteria/default/referrer.js

```diff
@@ -0,0 +1,15 @@
+import { setMatchingAttribute } from '../utils';
+
+const matchingAttribute = ( { store } ) => {
+	const value = document.referrer
+		? ( new URL( document.referrer ).hostname.replace( 'www.', '' ) || '' ).toLowerCase()
+		: '';
+	// Persist the referrer in the store.
+	if ( value ) {
+		store.set( 'referrer', value );
```

Comment:
I think we should avoid writing to the store as much as possible in the matching functions.

Isn't `referrer` something the store could provide by default, just like it does with `article_view` and other info? It looks like something generic enough to live there.

Reply:
Why avoid? It's embedded support to a powerful API. More advanced criteria will very likely benefit from this.

Reply:
For me it's a matter of separation of concerns.

On one side, the criteria checks if the current reader matches a specific situation. In order to do that, it READS info from the store and whatever other information needed. But it only reads.

On the other side the store is updated based on user interaction and other events that happen. This happens somewhere else.

I feel that if we start to write data to the store as we are checking for criteria things will get really confusing, with one criteria potentially stepping in the toes of the others... And it will also become hard to track what's making updates to the store.

Also, our Reader Data Library could be as complete as possible, offering global and valuable information about the current reader to all its consumers. Referrers feels like one of those things that could be part of the store by default...

(If you ask me, I still think that the whole criteria system should belong to the Data Library as well)

Reply:
Mind that this is the matching attribute, not the matching function.

I agree that the matching function is only reading data to determine a bool result.

The matching attribute is generating the data, which can either be a string, which runs `store.get( string )`, or a function, which will create the data. The data creation can be just stored data processing with a dynamic result, but that might not always be the case, and it's in the context of data creation to also allow storing it.

Reply:
> (If you ask me, I still think that the whole criteria system should belong to the Data Library as well)

I don't know what that means. Criteria have requirements to integrate with our UI, and segments and be performant when calculating matching. It leverages the store library as much as possible while still being able to dynamically process data when it needs.

Reply:
I moved the referrer store item logic to this PR: https://github.com/Automattic/newspack-plugin/pull/2541